website/code_issues.qmd

```diff
@@ -128,8 +128,91 @@ foo <- function(..., verbose = TRUE){
 
 Functions can print per default, like the example above, as long as the printing can be turned off (here, by setting `verbose = FALSE`).
 
+:::{.callout-note}
+
+`print()` and `cat()` are not the only functions which can write output onto the console. The issue described in the recipe, also applies to the use of other printing function like `writeLines()`. 
+If you are using loggers to document your functions' process, make sure that users can set their log level such that not messages are displayed.
+
+:::
+
 
 
 ---------------------------------------------------------
 
-# Change of options() and par() 
\ No newline at end of file
+# Change of Options, Graphical Parameters and Working Directory
+
+## Problem
+
+You are changing the `par()`, `options()` or `setwd()` in your functions, examples, demos or vignettes without resetting them.
+
+## Solution 
+
+Reset the changed options, in your functions by using [`on.exit()`](https://stat.ethz.ch/R-manual/R-devel/library/base/html/on.exit.html) or in examples, demos and vignettes with an additional line of code after the example.
+
+### Details
+
+Ideally, the user's options are not changed at all. If they really have to be altered, restoring the previous values of user options is mandatory for CRAN packages. There are different ways of resetting for functions, and examples, demos or vignettes which are recommended by CRAN. 
+
+Changing `par()`, `options()` or `setwd()` all invisibly return the previous values and therefore these can be stored in variables using the assignment operator `<-` and later be restored by calling the variable name as argument in the respective function. 
+
+
+**For functions**:
+
+When changing options inside one of your package functions, you can use [`on.exit()`](https://stat.ethz.ch/R-manual/R-devel/library/base/html/on.exit.html) for restoring. 
+
+```{r, eval = FALSE}
+
+foo <- function(x){
+  
+  # par():
+  oldpar <- par(mfrow = c(2,2))
+  on.exit(par(oldpar))
+  
+  # options():
+  oldop <- options(digits = 3)
+  on.exit(options(oldop))
+  
+  # setwd():
+  oldwd <- setwd(".")
+  on.exit(setwd(oldwd))
+  
+  # your code which requires a changed option
+  
+}
+
+
+
+```
+
+This will reset the `par()`, `options()` and `setwd()`. The use of [`on.exit()`](https://stat.ethz.ch/R-manual/R-devel/library/base/html/on.exit.html) makes it possible to restore options before exiting a function even if the function breaks. Therefore it needs to be called immediately after the option change within a function.  
```

Comment:
It is unclear why it needs to be called immediately after the option change, I don't think there isn't any requirement to do that, although it is recommended to avoid forgetting to do that.

Reply:
In general it does not really matter when you call `on.exit()`. 
There is a case though where it makes a difference. Consider this example:

> foo <- function(){
>   opar <- par(mfrow = c(2,2))
>   
>   on.exit(par(opar))
>   
>   stop("The function breaks here because of a bug")
>   
> }
> 
> 
> foo2 <- function(){
>   opar <- par(mfrow = c(2,2))
>   
>   stop("The function breaks here because of a bug")
>   
>   on.exit(par(opar))
>   
> }
> 
> foo()
> par("mfrow")
> 
> foo2()
> par("mfrow")


When the function breaks between the `par()` and `on.exit()` call, `on.exit()` is never called and thus the `par()` won't be restored.

Reply:
Totally, agree but you can have your on exit a couple of lines below after some other functions call that need that option set. I wanted to point out that such a strong language is misleading. Perhaps this example is also worth adding as a motivation on why it is important to add `on.exit()` soon.